File: flutter_qrscan_plugin-master/android/src/main/kotlin/dev/flutter/qrscan_plugin/QrScanPlatformView.kt
package dev.flutter.qrscan_plugin

import android.content.Context
import android.view.View
import androidx.camera.core.TorchState
import androidx.core.content.ContextCompat
import dev.flutter.qrscan_plugin.internal.QrScanListener
import dev.flutter.qrscan_plugin.internal.QrScanResult
import dev.flutter.qrscan_plugin.internal.QrScanView
import io.flutter.plugin.common.*
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class QrScanPlatformView(
    private val messenger: BinaryMessenger,
    private val context: Context,
    private val viewId: Int,
    private val args: Any?
) : PlatformView, MethodChannel.MethodCallHandler, EventChannel.StreamHandler, QrScanListener {

    companion object {
        private const val TAG = "QrScanPlatformView"
        const val VIEW_TYPE_ID = "dev.flutter.qrscan_plugin/QrScanPlatformView"
    }

    private val qrScanView = QrScanView(context)
    private val methodChannel = MethodChannel(messenger, "$VIEW_TYPE_ID/method#$viewId")
    private val eventChannel = EventChannel(messenger, "$VIEW_TYPE_ID/event#$viewId")
    private var eventSink: EventChannel.EventSink? = null

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
        qrScanView.setQrScanListener(this)
    }

    override fun getView(): View {
        qrScanView.didOnResumed()
        qrScanView.startCamera()
        return qrScanView
    }

    override fun dispose() {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        qrScanView.setQrScanListener(null)
        qrScanView.stopCamera()
        qrScanView.didOnDestroyed()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startCamera" -> handleStartCamera(result)
            "stopCamera" -> handleStopCamera(result)
            "enableTorch" -> handleEnableTorch(call, result)
            "setLinearZoom" -> handleSetLinearZoom(call, result)
            "getTorchState" -> handleGetTorchState(result)
            "setContinueAnalyze" -> handleSetContinueAnalyze(result)
        }
    }

    private fun handleStartCamera(result: MethodChannel.Result) {
        qrScanView.startCamera()
        result.success(true)
    }

    private fun handleStopCamera(result: MethodChannel.Result) {
        qrScanView.stopCamera()
        result.success(true)
    }

    private fun handleEnableTorch(call: MethodCall, result: MethodChannel.Result) {
        val future = qrScanView.cameraControl?.enableTorch(call.arguments as Boolean)
        if (future == null) {
            result.success(false)
        } else {
            future.addListener(Runnable {
                result.success(true)
            }, ContextCompat.getMainExecutor(context))
        }
    }

    private fun handleSetLinearZoom(call: MethodCall, result: MethodChannel.Result) {
        val linearZoom = (call.arguments as Double).toFloat()
        if (linearZoom !in 0.0..1.0) {
            result.success(false)
            return
        }
        val future = qrScanView.cameraControl?.setLinearZoom((call.arguments as Double).toFloat())
        if (future == null) {
            result.success(false)
        } else {
            future.addListener(Runnable {
                result.success(true)
            }, ContextCompat.getMainExecutor(context))
        }
    }

    private fun handleGetTorchState(result: MethodChannel.Result) {
        result.success(qrScanView.camera?.cameraInfo?.torchState?.value == TorchState.ON)
    }


    private fun handleSetContinueAnalyze(result: MethodChannel.Result) {
        qrScanView.setContinueAnalyze()
        result.success(false)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onFirstFrame() {
        super.onFirstFrame()
        val event = mapOf(
            "action" to "onFirstFrame"
        )
        eventSink?.success(event)
    }

    override fun onLuminosity(luminosity: Double) {
        super.onLuminosity(luminosity)
        val event = mapOf(
            "action" to "onLuminosity",
            "luminosity" to luminosity
        )
        eventSink?.success(event)
    }

    override fun onResult(result: QrScanResult) {
        super.onResult(result)
        val results = result.results.map {
            mapOf(
                "text" to it.text,
                "barcodeFormat" to it.barcodeFormat.name,
                "points" to it.resultPoints.map { p -> floatArrayOf(p.x, p.y) }
            )
        }
        val event = mapOf(
            "action" to "onResult",
            "previewJpeg" to result.previewJpeg,
            "previewSize" to intArrayOf(result.previewSize.width, result.previewSize.height),
            "results" to results
        )
        eventSink?.success(event)
    }


    class Factory(
        private val messenger: BinaryMessenger
    ) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
        override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
            return QrScanPlatformView(messenger, context, viewId, args)
        }
    }


}

File: flutter_qrscan_plugin-master/ios/Classes/QRScanPlatformView.swift
//
//  QRScanView.swift
//  flutter_qrscan_plugin
//
//  Created by SimMan on 2021/10/25.
//

import Foundation
import Flutter
import CoreMedia
import CoreImage
import ZXingObjC

class QRScanPlatformView: UIView, CaptureResultDelegate {
    private let channelName = "dev.flutter.qrscan_plugin/QrScanPlatformView"
    private let pluginChannelName = "dev.flutter.qrscan_plugin/FlutterQrScanPlugin"
    
    var binaryMessager: FlutterBinaryMessenger?
    private var _captureController: CaptureController?
    private var _channel: FlutterMethodChannel?
    private var _pluginChannel: FlutterMethodChannel?
    private var _eventChannel: FlutterEventChannel?
    private var _events: FlutterEventSink?
    
    private var _viewId: Int64?
    
    private var _isSendOnFrameAvailable: Bool = false
    
    init(frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?, binaryMessenger messenger: FlutterBinaryMessenger) {
        self.binaryMessager = messenger
        self._channel = FlutterMethodChannel(name: "\(channelName)/method#\(viewId)", binaryMessenger: binaryMessager!)
        self._pluginChannel = FlutterMethodChannel(name: pluginChannelName, binaryMessenger: binaryMessager!)
        self._viewId = viewId
        super.init(frame: frame)
        _channel?.setMethodCallHandler(handle(_:result:))
        _pluginChannel?.setMethodCallHandler(handle(_:result:))
        
        initializeCamera()
        
    }
   
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("flutter handle method: \(call.method)")
        switch call.method {
        case "getTorchState":
            result(NSNumber.init(value: self._captureController?.torchMode == .on))
        case "enableTorch":
            let enable = call.arguments as? Bool
            if (enable == true) {
                self._captureController?.torchMode = .on
            } else {
                self._captureController?.torchMode = .off
            }
            result(NSNumber.init(value: true));
        case "stopCamera":
            self._captureController?.stop()
        case "startCamera":
            self._isSendOnFrameAvailable = false
            self._captureController?.start()
        case "setContinueAnalyze":
            self._captureController?.canRecognition = true
        case "analyzeImage":
            analyzeImage(call, result: result)
        default:
            break
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        self._captureController?.videoPreviewLayer.frame = self.bounds;
    }
    
    func initializeCamera() {
        self._captureController = CaptureController.init()
        _captureController?.metadataObjectsDelegate = self
        _captureController?.metadataObjectTypes = [.qr, .code39, .aztec, .code128, .code93, .dataMatrix, .ean13, .ean8]
        self.layer.addSublayer(_captureController!.videoPreviewLayer)
        _captureController?.onFrameAvailable = {
            if (!self._isSendOnFrameAvailable) {
                self._isSendOnFrameAvailable = true
                self._events?(["action": "onFirstFrame"])
            }
        }
        self._eventChannel = FlutterEventChannel(name: "\(channelName)/event#\(_viewId!)", binaryMessenger: binaryMessager!)
        self._eventChannel?.setStreamHandler(self)
        _captureController?.start()
    }
    
    func metadataOutputObjects(results: [QRResult], image: UIImage) {
        self._captureController?.canRecognition = false
        let result = [
            "action": "onResult",
            "previewJpeg": FlutterStandardTypedData.init(bytes: image.jpegData(compressionQuality: 90)!),
            "previewSize": [Float(self._captureController!.videoPreviewLayer.frame.size.width), Float(self._captureController!.videoPreviewLayer.frame.size.height)],
            "results": results.map { $0.toMap() }
        ] as [String : Any]
        self._events?(result);
        
    }
    
    func onBrightnessValueChange(value: CGFloat) {
        self._events?(["action": "onLuminosity", "luminosity": Float.init(value)]);
    }
    
    func analyzeImage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let image = UIImage.init(contentsOfFile: call.arguments as! String), let cgImage = image.cgImage else { return }

        let ciImage = CIImage(cgImage: cgImage)
        let detector = CIDetector(ofType: CIDetectorTypeQRCode, context: CIContext(options: nil), options: [CIDetectorAccuracy: CIDetectorAccuracyHigh])
        let features = detector?.features(in: ciImage)
        if let feature = features?.first as? CIQRCodeFeature, let code = feature.messageString {
            let points: [[Double]] = [
                [Double(feature.topLeft.x), Double(feature.topLeft.y)],
                [Double(feature.topRight.x), Double(feature.topRight.y)],
                [Double(feature.bottomLeft.x), Double(feature.bottomLeft.y)],
                [Double(feature.bottomRight.x), Double(feature.bottomRight.y)],
            ];
            result([
                "text": code,
                "barcodeFormat": "QR_CODE",
                "points": points
            ])
            return
        }
    
        guard let fixOrientImage: UIImage = UIImage.fixedOrientation(for: image) else { return }
        let scaledImage = QRScanUtils.scaledImage(fixOrientImage, maxWidth: 1080, maxHeight: 1080)
        
        let source = ZXCGImageLuminanceSource(cgImage: scaledImage.cgImage)
        if let binary = ZXHybridBinarizer.init(source: source!) {
            let bitmap = ZXBinaryBitmap.init(binarizer: binary)
            let hints = ZXDecodeHints.hints() as! ZXDecodeHints
            let reader = ZXMultiFormatReader.reader() as! ZXReader
            // let multipleReader = ZXGenericMultipleBarcodeReader(delegate: reader)

            do {
                // let results = try multipleReader?.decodeMultiple(bitmap, hints: hints)
                let _zxResult = try reader.decode(bitmap, hints: hints) as ZXResult
                let _zxResultPoints = _zxResult.resultPoints as! [ZXResultPoint]
                let points = _zxResultPoints.map { $0.pointArray() }
                
                result([
                    "text": _zxResult.text ?? "",
                    "barcodeFormat": _zxResult.barcodeFormat.stringValue(),
                    "points": points
                ])
            } catch {
                print("\(error)")
                result(nil)
            }
        }
    }
}

// MARK: - FlutterStreamHandler
extension QRScanPlatformView: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self._events = events;
        return nil;
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self._events = nil;
        return nil;
    }
}
